index.js

```diff
@@ -1,79 +1,45 @@
 'use strict';
 
-const bigquery = require('./lib/bigquery');
+const logger = require('./lib/logger');
 const Inputs = require('./lib/inputs');
 
 exports.handler = (event, context, callback) => {
-  let errs = [];
-  function concatErrors() {
-    errs.forEach(e => exports.logError(`${e}`));
-    if (errs.length === 1) {
-      return errs[0];
-    } else if (errs.length > 1) {
-      let msgs = errs.map(e => `${e}`).join(', ');
-      return new Error(`Multiple errors: ${msgs}`);
-    } else {
-      return null;
-    }
-  }
+  let fatalErr, records;
 
   // decode the base64 kinesis records
   if (!event || !event.Records) {
-    let err = new Error(`Invalid event input: ${JSON.stringify(event)}`);
-    exports.logError(`${err}`);
-    return callback(err);
+    fatalErr = `Invalid event input: ${JSON.stringify(event)}`;
+  } else {
+    records = event.Records.map(r => {
+      try {
+        return JSON.parse(new Buffer(r.kinesis.data, 'base64').toString('utf-8'));
+      } catch (decodeErr) {
+        fatalErr = `Invalid record input: ${JSON.stringify(r)}`;
+      }
+    });
   }
-  let records = event.Records.map(r => {
-    try {
-      return JSON.parse(new Buffer(r.kinesis.data, 'base64').toString('utf-8'));
-    } catch (err) {
-      errs.push(new Error(`Invalid record input: ${JSON.stringify(r)}`));
-      return false;
-    }
-  }).filter(r => r);
 
-  // input records and check for unrecognized
-  let inputs = new Inputs(records);
-  inputs.unrecognized.forEach(r => {
-    errs.push(new Error(`Unrecognized input record: ${JSON.stringify(r)}`));
-  });
+  // NOTE: callback with errors here will cause the lambda to run indefinitely
+  // on the same kinesis records... so you'd better monitor for errors.
```

Comment:
Well, eventually the kinesis records will time out at 24 hours.  But as far as I can tell from the [Stream-based event sources](http://docs.aws.amazon.com/lambda/latest/dg/retries-on-errors.html) docs, it appears failures will block by design.

So we'd need some sort of "state" tracker, for the next run to know about the last run on this batch of records.  Since we can't modify the kinesis data, I'm guessing it would have to be redis or some datastore?  I don't see any forum posts of people doing this.

But I do see that lambdas will automagically "back off" after a number of failures ... eventually settling on 1-min between retries.  Which is good news for our wallet!

Reply:
that sounds reasonable - not looking to add a "state" to this